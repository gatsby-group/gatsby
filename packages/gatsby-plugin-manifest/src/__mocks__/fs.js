const fs = jest.genMockFromModule(`fs`)

fs.__setWriteFileSyncFn = fn => (fs.writeFileSync = fn)

module.exports = fs
