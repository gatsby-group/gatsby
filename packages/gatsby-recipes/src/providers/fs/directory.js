const fs = require(`fs-extra`)
const path = require(`path`)
const Joi = require(`@hapi/joi`)
const isBinaryPath = require(`is-binary-path`)

const getDiff = require(`../utils/get-diff`)
const resourceSchema = require(`../resource-schema`)

const makePath = (root, relativePath) => path.join(root, relativePath)

const directoryExists = fullPath => {
  try {
    fs.accessSync(fullPath, fs.constants.F_OK)
    return true
  } catch (e) {
    return false
  }
}

const create = async ({ root }, { id, path: directoryPath }) => {
  const fullPath = makePath(root, directoryPath)
  await fs.ensureDir(fullPath)
  return await read({ root }, directoryPath)
}

const update = async (context, resource) => {
  const fullPath = makePath(context.root, resource.id)
  await fs.ensureDir(fullPath)
  return await read(context, resource.id)
}

const read = async (context, id) => {
  const fullPath = makePath(context.root, id)

  if (!directoryExists(fullPath)) {
    return undefined
  }

  const resource = { id, path: id }
  resource._message = message(resource)
  return resource
}

const destroy = async (context, directoryResource) => {
  const fullPath = makePath(context.root, directoryResource.id)
  await fs.rmdirSync(fullPath)
  return directoryResource
}

// TODO pass action to plan
module.exports.plan = async (context, { id, path: directoryPath }) => {
  let currentResource
  if (!isBinaryPath(directoryPath)) {
    currentResource = await read(context, directoryPath)
  } else {
    currentResource = `Binary file`
  }

  let newState = directoryPath

  const plan = {
    currentState: (currentResource && currentResource.content) || ``,
    newState,
    describe: `Create ${directoryPath}`,
    diff: ``,
  }

  if (plan.currentState !== plan.newState) {
    plan.diff = await getDiff(plan.currentState, plan.newState)
  }

  return plan
}

const message = resource => `Created directory ${resource.path}`

const schema = {
  path: Joi.string(),
  content: Joi.string(),
  ...resourceSchema,
}
exports.schema = schema
exports.validate = resource =>
  Joi.validate(resource, schema, { abortEarly: false })

module.exports.create = create
module.exports.update = update
module.exports.read = read
module.exports.destroy = destroy
