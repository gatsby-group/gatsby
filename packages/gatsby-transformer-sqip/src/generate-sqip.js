const createContentDigest = require(`../../gatsby/src/utils/create-content-digest`)
const { resolve, parse } = require(`path`)

const Debug = require(`debug`)
const { exists, readFile, writeFile } = require(`fs-extra`)
const svgToMiniDataURI = require(`mini-svg-data-uri`)
const PQueue = require(`p-queue`)
const sqip = require(`sqip`)

const queue = new PQueue({ concurrency: 1 })
const debug = Debug(`gatsby-transformer-sqip`)

module.exports = async function generateSqip(options) {
  const {
    cache,
    absolutePath,
    numberOfPrimitives,
    blur,
    mode,
    cacheDir,
  } = options

  debug({ options })

  const { name } = parse(absolutePath)

  const sqipOptions = {
    numberOfPrimitives,
    blur,
    mode,
  }

  const optionsHash = createContentDigest(JSON.stringify(sqipOptions))

  const cacheKey = `sqip-${name}-${optionsHash}`
  const cachePath = resolve(cacheDir, `${name}-${optionsHash}.svg`)
  let primitiveData = await cache.get(cacheKey)

  debug({ primitiveData })

  if (!primitiveData) {
    let svg
    if (await exists(cachePath)) {
      const svgBuffer = await readFile(cachePath)
      svg = svgBuffer.toString()
    } else {
      debug(`generate sqip for ${name}`)
      const result = await queue.add(
        async () =>
          new Promise((resolve, reject) => {
            try {
              const result = sqip({
                filename: absolutePath,
                ...sqipOptions,
              })
              resolve(result)
            } catch (error) {
              reject(error)
            }
          })
      )

      svg = result.final_svg

      await writeFile(cachePath, svg)
    }

    primitiveData = {
      svg,
      dataURI: svgToMiniDataURI(svg),
    }

    await cache.set(cacheKey, primitiveData)
  }

  return primitiveData
}
