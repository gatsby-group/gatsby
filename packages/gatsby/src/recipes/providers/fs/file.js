const fs = require(`fs-extra`)
const path = require(`path`)
const mkdirp = require(`mkdirp`)
const Joi = require(`@hapi/joi`)

const fileExists = fullPath => {
  try {
    fs.accessSync(fullPath, fs.constants.F_OK)
    return true
  } catch (e) {
    return false
  }
}

const create = async ({ root }, { id, path: filePath, content }) => {
  const fullPath = path.join(root, filePath)
  const { dir } = path.parse(fullPath)

  await mkdirp(dir)

  await fs.writeFile(fullPath, content)

  return await read({ root }, fullPath)
}

const update = async (context, resource) => {
  await fs.writeFile(resource.id, resource.content)
  return await read(context, resource.id)
}

const read = async (context, id) => {
  let content = ``
  if (fileExists(id)) {
    content = await fs.readFile(id, `utf8`)
  }

  return { id, path: id, content }
}

const destroy = async (context, fileResource) => {
  await fs.unlink(fileResource.path)
}

// TODO pass action to plan
module.exports.plan = async (context, { id, path: filePath, content }) => {
  let fullPath
  if (!id) {
    fullPath = path.join(context.root, filePath)
  } else {
    fullPath = id
  }
  const currentResource = await read(context, fullPath)

  return {
    currentState: currentResource.content,
    newState: content,
    describe: `Write ${fullPath}`,
  }
}

exports.validate = () => {
  return {
    path: Joi.string(),
    content: Joi.string(),
  }
}

module.exports.exists = fileExists

module.exports.create = create
module.exports.update = update
module.exports.read = read
module.exports.destroy = destroy
