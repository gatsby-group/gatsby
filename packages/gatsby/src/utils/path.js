const path = require(`path`)
const { joinPath, createContentDigest } = require(`gatsby-core-utils`)

export function withBasePath(basePath) {
  return (...paths) => joinPath(basePath, ...paths)
}

export function withTrailingSlash(basePath) {
  return `${basePath}/`
}

const posixJoinWithLeadingSlash = paths =>
  path.posix.join(
    ...paths.map((segment, index) =>
      segment === `` && index === 0 ? `/` : segment
    )
  )

export function getCommonDir(path1, path2) {
  const path1Segments = path1.split(/[/\\]/)
  const path2Segments = path2.split(/[/\\]/)

  for (let i = 0; i < path1Segments.length; i++) {
    if (i >= path2Segments.length) {
      return posixJoinWithLeadingSlash(path2Segments)
    } else if (
      path1Segments[i].toLowerCase() !== path2Segments[i].toLowerCase()
    ) {
      const joined = path1Segments.slice(0, i)
      return posixJoinWithLeadingSlash(joined)
    }
  }

  return posixJoinWithLeadingSlash(path1Segments)
}

const MAX_PATH_SEGMENT_CHARS = 255
const MAX_PATH_SEGMENT_BYTES = 255
const SLICING_INDEX = 100
const SLICING_INDEX_BYTES = 50
const pathSegmentRe = /[^/]+/g

const isMacOs = process.platform === `darwin`
const isWindows = process.platform === `win32`

export const truncatePath = path =>
  path.replace(pathSegmentRe, match => {
    if (isMacOS || isWindows) {
      if (match.length > MAX_PATH_SEGMENT_CHARS) {
        return (
          match.slice(0, SLICING_INDEX) +
          createContentDigest(match.slice(SLICING_INDEX))
        )
      }
    } else {
      if (Buffer.from(match).length > MAX_PATH_SEGMENT_BYTES) {
        return (
          match.slice(0, SLICING_INDEX_BYTES) +
          createContentDigest(match.slice(SLICING_INDEX_BYTES))
        )
      }
    }

    return match
  })
