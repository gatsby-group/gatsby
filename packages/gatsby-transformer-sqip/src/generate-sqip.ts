import { resolve, parse } from "path"

import Debug from "debug"
import { promises, constants } from "fs"
import PQueue from "p-queue"
import { sqip, SqipImageMetadata, SqipPluginOptions, SqipResult } from "sqip"
import { createContentDigest } from "gatsby-core-utils"
import { GatsbyCache } from "gatsby"

const queue = new PQueue({ concurrency: 1 })
const debug = Debug(`gatsby-transformer-sqip`)

export async function generateSqip(options: {
  cache: GatsbyCache
  absolutePath: string
  numberOfPrimitives?: number
  blur?: number
  mode?: number
  cacheDir: string
  contentDigest: string
}): Promise<{
  svg: string
  dataURI: string
  metadata: SqipImageMetadata
  additionalMetadata: { [key: string]: unknown }
} | null> {
  const {
    cache,
    absolutePath,
    numberOfPrimitives,
    blur,
    mode,
    cacheDir,
    contentDigest,
  } = options

  debug({ options })

  const { name, ext } = parse(absolutePath)

  if (!ext.match(/(jpe?g|png|gif)$/)) {
    debug(`Unsupported file type ${name} (${contentDigest})`)
    return null
  }

  const sqipBaseConfig: Omit<SqipPluginOptions["sqipConfig"], "input"> = {
    plugins: [
      {
        name: `sqip-plugin-primitive`,
        options: {
          numberOfPrimitives,
          mode,
        },
      },
      {
        name: `sqip-plugin-blur`,
        options: {
          blur,
        },
      },
      `sqip-plugin-svgo`,
      `sqip-plugin-data-uri`,
    ],
  }

  const optionsHash = createContentDigest(sqipBaseConfig)

  const cacheKey = `${contentDigest}-${optionsHash}`
  const cachePathSvg = resolve(cacheDir, `${contentDigest}-${optionsHash}.svg`)
  const cachePathJson = resolve(
    cacheDir,
    `${contentDigest}-${optionsHash}.json`
  )

  debug(
    `Request preview generation for ${name} (${contentDigest}-${optionsHash})`
  )

  return queue.add(async () => {
    let primitiveData = await cache.get(cacheKey)
    let svg: Buffer
    let meta: SqipImageMetadata
    // let dataURI: string
    if (!primitiveData) {
      debug(
        `Executing preview generation request for ${name} (${contentDigest}-${optionsHash})`
      )

      try {
        try {
          await promises.access(cachePathSvg, constants.R_OK)
          await promises.access(cachePathJson, constants.R_OK)
          debug(
            `SQIP data already generated for ${name} (${contentDigest}-${optionsHash})`
          )
          svg = await promises.readFile(cachePathSvg)
          const metaRes = await promises.readFile(cachePathJson)
          meta = JSON.parse(metaRes.toString())
        } catch {
          debug(
            `Generateing SQIP data for ${name} (${contentDigest}-${optionsHash})`
          )

          const res = await sqip({
            input: absolutePath,
            ...sqipBaseConfig,
          })

          const result: SqipResult = res as SqipResult

          svg = result.content
          meta = result.metadata

          await promises.writeFile(cachePathSvg, svg)
          await promises.writeFile(cachePathJson, JSON.stringify(meta))
          debug(
            `Wrote primitive result file to disk for ${name} (${contentDigest}-${optionsHash})`
          )
        }

        // Enhance palette with color hex code
        const enhancedPalette = {}
        for (const k of Object.keys(meta.palette)) {
          enhancedPalette[k] = {
            population: meta.palette[k]?.population,
            rgb: meta.palette[k]?.rgb,
            hex: meta.palette[k]?.hex,
            hsl: meta.palette[k]?.hsl,
          }
        }

        // Split up metadata in default and additional
        const metadata = {}
        const additionalMetadata = {}
        for (const [k, v] of Object.entries(meta)) {
          if (k === `palette`) {
            metadata[k] = enhancedPalette
          } else if (
            [
              `originalWidth`,
              `originalHeight`,
              `type`,
              `width`,
              `height`,
            ].includes(k)
          ) {
            metadata[k] = v
          } else {
            additionalMetadata[k] = v
          }
        }

        primitiveData = {
          svg: svg.toString(),
          metadata,
          additionalMetadata: additionalMetadata,
        }

        await cache.set(cacheKey, primitiveData)
      } catch (err) {
        err.message = `Unable to generate SQIP for ${name} (${contentDigest}-${optionsHash})\n${err.message}`

        throw err
      }
    }

    return primitiveData
  })
}
