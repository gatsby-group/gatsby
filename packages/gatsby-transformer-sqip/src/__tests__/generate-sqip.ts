import { resolve } from "path"

import { promises } from "fs"
import { sqip } from "sqip"

import { generateSqip } from "../generate-sqip"

jest.mock(`sqip`, () => {
  const originalModule = jest.requireActual(`sqip`)
  return {
    ...originalModule,
    sqip: jest.fn(() => {
      return {
        content: Buffer.from(`<svg><!-- Mocked SQIP SVG --></svg>`),
        metadata: {
          palette: {
            Vibrant: {
              rgb: [1, 2, 3],
              hsl: [0.1, 0.2, 0.3],
              hex: `#M0CK3D`,
              population: 42,
            },
          },
          dataURI: `data:mocked`,
        },
      }
    }),
  }
})

jest.mock(`fs`, () => {
  const originalModule = jest.requireActual(`fs`)
  return {
    ...originalModule,
    promises: {
      ...originalModule.promises,
      // access: access,
      access: jest.fn(() => {
        throw new Error(`file does not exist`)
      }),
      readFile: jest.fn(() =>
        Buffer.from(`<svg><!-- Cached SQIP SVG --></svg>`)
      ),
      writeFile: jest.fn(),
    },
  }
})

afterEach(() => {
  sqip.mockClear()
  promises.access.mockClear()
  promises.readFile.mockClear()
  promises.writeFile.mockClear()
})

describe(`gatsby-transformer-sqip`, () => {
  const absolutePath = resolve(
    __dirname,
    `images`,
    `this-file-does-not-need-to-exist-for-the-test.jpg`
  )
  const cacheDir = __dirname

  describe(`generateSqip`, () => {
    it(`not cached`, async () => {
      const cache = {
        get: jest.fn(),
        set: jest.fn(),
      }
      const numberOfPrimitives = 5
      const blur = 0
      const mode = 3
      const result = await generateSqip({
        cache,
        cacheDir,
        absolutePath,
        numberOfPrimitives,
        blur,
        mode,
      })
      expect(result).toMatchSnapshot()
      expect(sqip).toHaveBeenCalledTimes(1)
      const sqipArgs = sqip.mock.calls[0][0]
      expect(sqipArgs.input).toMatch(absolutePath)
      delete sqipArgs.input
      expect(sqipArgs).toMatchSnapshot()

      expect(promises.access).toHaveBeenCalledTimes(1)
      expect(promises.writeFile).toHaveBeenCalledTimes(2)
      expect(promises.readFile).toHaveBeenCalledTimes(0)
    })
    it(`cached`, async () => {
      promises.access.mockImplementation(() => true)
      promises.readFile.mockImplementationOnce(() =>
        Buffer.from(`<svg><!-- Cached SQIP SVG --></svg>`)
      )
      promises.readFile.mockImplementationOnce(() =>
        Buffer.from(`{"mocked": "json", "palette": {}}`)
      )
      const cache = {
        get: jest.fn(),
        set: jest.fn(),
      }
      const numberOfPrimitives = 5
      const blur = 0
      const mode = 3
      const result = await generateSqip({
        cache,
        cacheDir,
        absolutePath,
        numberOfPrimitives,
        blur,
        mode,
      })

      expect(result).toMatchSnapshot()

      expect(promises.access).toHaveBeenCalledTimes(2)
      expect(promises.writeFile).toHaveBeenCalledTimes(0)
      expect(promises.readFile).toHaveBeenCalledTimes(2)
      expect(sqip).toHaveBeenCalledTimes(0)
    })
    it(`returns null for unsupported files`, async () => {
      promises.access.mockImplementationOnce(() => true)

      const cache = {
        get: jest.fn(),
        set: jest.fn(),
      }
      const numberOfPrimitives = 5
      const blur = 0
      const mode = 3
      const result = await generateSqip({
        cache,
        cacheDir,
        absolutePath: absolutePath.replace(`.jpg`, `.svg`),
        numberOfPrimitives,
        blur,
        mode,
      })

      expect(result).toBe(null)

      expect(sqip).toHaveBeenCalledTimes(0)
      expect(promises.access).toHaveBeenCalledTimes(0)
      expect(promises.writeFile).toHaveBeenCalledTimes(0)
      expect(promises.readFile).toHaveBeenCalledTimes(0)
    })
  })
})
