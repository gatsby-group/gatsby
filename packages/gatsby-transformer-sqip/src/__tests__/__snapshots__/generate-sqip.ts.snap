// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-transformer-sqip generateSqip cached 1`] = `
Object {
  "additionalMetadata": Object {
    "mocked": "json",
  },
  "metadata": Object {
    "palette": Object {},
  },
  "svg": "<svg><!-- Cached SQIP SVG --></svg>",
}
`;

exports[`gatsby-transformer-sqip generateSqip not cached 1`] = `
Object {
  "additionalMetadata": Object {
    "dataURI": "data:mocked",
  },
  "metadata": Object {
    "palette": Object {
      "Vibrant": Object {
        "hex": "#M0CK3D",
        "hsl": Array [
          0.1,
          0.2,
          0.3,
        ],
        "population": 42,
        "rgb": Array [
          1,
          2,
          3,
        ],
      },
    },
  },
  "svg": "<svg><!-- Mocked SQIP SVG --></svg>",
}
`;

exports[`gatsby-transformer-sqip generateSqip not cached 2`] = `
Object {
  "plugins": Array [
    Object {
      "name": "sqip-plugin-primitive",
      "options": Object {
        "mode": 3,
        "numberOfPrimitives": 5,
      },
    },
    Object {
      "name": "sqip-plugin-blur",
      "options": Object {
        "blur": 0,
      },
    },
    "sqip-plugin-svgo",
    "sqip-plugin-data-uri",
  ],
}
`;
